src/snap_lock.py

```diff
@@ -0,0 +1,119 @@
+"""Snap Lock Module using fcntl.flock and file-based reference counting."""
+
+import os
+import re
+import time
+from contextlib import contextmanager
+from typing import Generator
+import errno
+
+
+class SnapLockError(Exception):
+    """Base class for SnapLock error."""
+
+    pass
+
+
+class SnapLock:
+    """SnapLock provides mechanisms for singleton snaps using fcntl.flock and file-based reference counting.
+
+    Raises SnapLockError on any error.
+    """
+
+    def __init__(self, instance_name: str, lock_timeout_sec: int = 300):
+        # Replace all non-alphanumeric chars (except _-) with _
+        self.instance_name = re.sub(r'[^\w-]', '_', instance_name)
+        self.lock_timeout = lock_timeout_sec
+        self.separator = '--'
+
+        self.lock_dir = os.path.join('/run/lock', 'snap_locks')
+        try:
+            os.makedirs(self.lock_dir, exist_ok=True)
+            os.chown(self.lock_dir, os.geteuid(), os.getegid())
+        except OSError as e:
+            if e.errno != errno.EEXIST:
+                raise SnapLockError(f'Error creating lock directory: {e}') from e
+
+    def _get_lock_file_path(self, resource_name: str) -> str:
+        return os.path.join(self.lock_dir, f'{resource_name}.lock')
```

Comment:
Could we narrow down the terminology? Currently we have
- lock dir and lock file
- resource name
- snap name
- reference file
- instance name

I am confused :)